fda/grid.py

```diff
@@ -677,92 +685,157 @@ def concatenate(self, other):
+        if fig is not None and ax is not None:
+            raise ValueError("fig and axes parameters cannot be passed as arguments at the same time.")
+
+        if fig != None and len(fig.get_axes()) != self.ndim_image:
+            raise ValueError("Number of axes of the figure must be equal to"
+                             "the dimension of the image.")
+
+        if ax is not None and len(ax)!= self.ndim_image:
+            raise ValueError("Number of axes must be equal to the dimension of the image.")
+
+        if fig == None and ax == None:
+            fig, ax = self.set_figure_and_axes()
+
+        if self.ndim_domain == 1:
+            for i in range(self.ndim_image):
+                ax[i].plot(self.sample_points[0], self.data_matrix[:, :, i].T, **kwargs)
+        else:
+            X = self.sample_points[0]
+            Y = self.sample_points[1]
+            X, Y = numpy.meshgrid(X, Y)
+            for i in range(self.ndim_image):
+                for j in range(self.nsamples):
+                    ax[i].plot_surface(X, Y, numpy.squeeze(self.data_matrix[j, :, :, i]).T, **kwargs)
+
+        self.set_labels(fig, ax)
+
+        return fig, ax
+
+    def scatter(self, fig=None, ax = None, **kwargs):
         """Scatter plot of the FDatGrid object.
 
         Args:
-            ax (axis object, optional): axis over with the graphs are plotted.
-                Defaults to matplotlib current axis.
-            **kwargs: keyword arguments to be passed to the
-                matplotlib.pyplot.scatter function.
+            fig (figure object, optional): figure over with the graphs are plotted in case ax is not specified.
+                If None and ax is also None, the figure is initialized.
+            ax (list of axis objects, optional): axis over where the graphs are plotted. If None, see param fig.
+            **kwargs: keyword arguments to be passed to the matplotlib.pyplot.scatter function;
 
         Returns:
-            :obj:`matplotlib.collections.PathCollection`
+            fig (figure object): figure object in which the graphs are plotted in case ax is None.
+            ax (axes object): axes in which the graphs are plotted.
 
         """
-        if self.ndim_domain != 1:
-            raise NotImplementedError("Scatter only supported for functional "
-                                      "data with a domain dimension of 1.")
-
-        if self.ndim_image != 1:
-            raise NotImplementedError("Scatter only supported for functional "
-                                      "data with a image dimension of 1.")
-        if ax is None:
-            ax = matplotlib.pyplot.gca()
-        _plot = None
-        for i in range(self.nsamples):
-            _plot = ax.scatter(self.sample_points[0],
-                               self.data_matrix[i],
-                               **kwargs)
-        self._set_labels(ax)
-        return _plot
+        if self.ndim_domain > 2:
```

Comment:
Not very important, but maybe the checks can be extracted to a function, as they are used twice (in `plot` and `scatter`).

In fact, the two functions are nearly identical, so maybe you can use a private function and parameterize the parts that are not common.

Reply:
unification of checks done